File: node_ws/src/demo_python_pkg/demo_python_pkg/person_node.py

import rclpy
from rclpy.node import Node
"""
PersonNode：人类节点
继承 Node 类，其属性和方法均被继承
"""
class PersonNode(Node):
    def __init__(self, node_name: str, name: str, age: int):
        """
        初始化阶段：使用super().__init__()
        运行阶段：使用self.method()
        """
        super().__init__(node_name)
        # 类属性
        self.name = name
        self.age = age

    def eat(self, food_name: str):
        """
        方法：吃东西
        :food_name 食物名称
        """
        # print(f"{self.name}，{self.age}岁，爱吃{food_name}")
        
        # 调用Node 类中的get_logger方法，并打印日志
        self.get_logger().info(f"{self.name}，{self.age}岁，爱吃{food_name}")
    

def main():
    rclpy.init()
    node = PersonNode('zhangsan', '法外狂徒张三', 10)
    # node1 = PersonNode('法外狂徒王五', 15)
    node.eat('鱼香肉丝')
    # node1.eat('排骨')
    rclpy.spin(node)
    rclpy.shutdown()

File: node_ws/src/demo_cpp_pkg/src/person_node.cpp
// 双引号：现在项目目录查找，再在系统目录查找
#include "rclcpp/rclcpp.hpp"

// 尖括号：直接在系统目录查找
// #include <iostream>

/**
 * C++ 继承外部包中的类
 */
class PersonNode : public rclcpp::Node
{
    
private:
    // 私有属性
    std::string name_;
    int age_;

public:
    // 构造函数
    /* &var 引用传递，避免拷贝 */
    PersonNode(const std::string &node_name, const std::string &name, const int &age)
        : Node(node_name) /* 调用父类的构造函数，等同于Python中的super().__init__*/
    {
        this->name_ = name;
        this->age_ = age;
    };
    // 方法
    void eat(const std::string &food_name)
    {
        // 宏定义的 RCLCPP_INFO
        RCLCPP_INFO(this->get_logger(), "我是%s, %d岁，爱吃%s", this->name_.c_str(), this->age_, food_name.c_str());
    };
};

int main(int argc, char** argv) // argc：命令行启动时传入的参数数量 argv：所有传参二维数组
{
    rclcpp::init(argc, argv); // 初始化

    // 智能指针 shared_ptr：基于引用计数
    auto node = std::make_shared<PersonNode>("person_node", "里斯", 18); // 创建节点
    RCLCPP_INFO(node->get_logger(), "你好 person_node");
    node->eat("鱼香肉丝");
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}